File: nixos/modules/programs/command-not-found/src/command-not-found.cpp
#include <memory>
#include <stdio.h>
#include <functional>
#include <string>
#include <vector>

#include <sqlite3.h>

int queryPackages(std::string system, std::string program,
                  std::vector<std::string> &packages) {
  sqlite3 *raw_db = nullptr;
  if (sqlite3_open(DB_PATH, &raw_db) != SQLITE_OK) {
    fprintf(stderr, "cannot open database \"%s\": %s\n", DB_PATH,
            sqlite3_errmsg(raw_db));
  };
  std::unique_ptr<sqlite3, std::function<void(sqlite3*)>> db(raw_db, [](auto db) { sqlite3_close(db); });

  const char *sql =
      "select package from Programs where system = ? and name = ?;";

  sqlite3_stmt *raw_stmt = nullptr;
  if (sqlite3_prepare_v2(db.get(), sql, -1, &raw_stmt, NULL) != SQLITE_OK) {
    fprintf(stderr, "Failed to prepare query: %s\n", sqlite3_errmsg(db.get()));
    return 1;
  }
  std::unique_ptr<sqlite3_stmt, std::function<void(sqlite3_stmt*)>> stmt(raw_stmt, [](auto stmt) { sqlite3_finalize(stmt); });

  if (sqlite3_bind_text(stmt.get(), 1, system.c_str(), -1, 0) != SQLITE_OK) {
    fprintf(stderr, "Failed to bind system parameter: %s\n",
            sqlite3_errmsg(db.get()));
    return 1;
  };
  if (sqlite3_bind_text(stmt.get(), 2, program.c_str(), -1, 0) != SQLITE_OK) {
    fprintf(stderr, "Failed to bind program parameter: %s\n",
            sqlite3_errmsg(db.get()));
    return 1;
  };

  while (sqlite3_step(stmt.get()) == SQLITE_ROW) {
    packages.push_back((const char *)(sqlite3_column_text(stmt.get(), 0)));
  }

  return 0;
}

int main(int argc, char **argv) {
  const char *system = getenv("NIX_SYSTEM");
  const char *program = argv[1];

  if (!system) {
    system = NIX_SYSTEM;
  }

  if (argc < 2) {
    fprintf(stderr, "USAGE: %s PROGRAM\n", argv[0]);
    return 1;
  }

  std::vector<std::string> packages = {};
  if (queryPackages(system, program, packages) < 0) {
    return 1;
  };

  if (packages.size()) {
    auto advice = packages.size() > 1 ?
      "It is provided by several packages. You can install it by typing on of the of following commands:\n" :
      "You can install it by typing:";
    fprintf(stderr,
            "The program '%s' is currently not installed. %s\n",
            program, advice);
    for (auto pkg : packages) {
      fprintf(stderr, "  nix-env -iA nixos.%s\n", pkg.c_str());
    }
  } else {
    fprintf(stderr, "%s: command not found\n", program);
  }

  return 127;
}

File: nixos/modules/programs/command-not-found/command-not-found.pl
#! @perl@/bin/perl -w

use strict;
use DBI;
use DBD::SQLite;
use String::ShellQuote;
use Config;

my $program = $ARGV[0];

my $dbPath = "@dbPath@";

my $dbh = DBI->connect("dbi:SQLite:dbname=$dbPath", "", "")
    or die "cannot open database `$dbPath'";
$dbh->{RaiseError} = 0;
$dbh->{PrintError} = 0;

my $system = $ENV{"NIX_SYSTEM"} // $Config{myarchname};

my $res = $dbh->selectall_arrayref(
    "select package from Programs where system = ? and name = ?",
    { Slice => {} }, $system, $program);

my $len = !defined $res ? 0 : scalar @$res;

if ($len == 0) {
    print STDERR "$program: command not found\n";
} elsif ($len == 1) {
    my $package = @$res[0]->{package};
    if ($ENV{"NIX_AUTO_RUN"} // "") {
        if ($ENV{"NIX_AUTO_RUN_INTERACTIVE"} // "") {
            while (1) {
                print STDERR "'$program' from package '$package' will be run, confirm? [yn]: ";
                chomp(my $comfirm = <STDIN>);
                if (lc $comfirm eq "n") {
                    exit 0;
                } elsif (lc $comfirm eq "y") {
                    last;
                }
            }
        }
		if ($ENV{"NIX_AUTO_RUN_FLAKE"} // "") {
			# relies on flake registry
			exec("nix", "shell", "nixpkgs#" . $package, "--command", shell_quote(@ARGV));
		} else {
			exec("nix-shell", "-p", $package, "--run", shell_quote("exec", @ARGV));
		}
    } else {
        print STDERR <<EOF;
The program '$program' is not in your PATH. You can make it available in an
ephemeral shell by typing:
  nix-shell -p $package
EOF
    }
} else {
    if ($ENV{"NIX_AUTO_RUN"} // "") {
        print STDERR "Select a package that provides '$program':\n";
        for my $i (0 .. $len - 1) {
            print STDERR "  [", $i + 1, "]: @$res[$i]->{package}\n";
        }
        my $choice = 0;
        while (1) { # exec will break this loop
            no warnings "numeric";
            print STDERR "Your choice [1-${len}]: ";
            # 0 can be invalid user input like non-number string
            # so we start from 1
            $choice = <STDIN> + 0;
            if (1 <= $choice && $choice <= $len) {
				if ($ENV{"NIX_AUTO_RUN_FLAKE"} // "") {
					# relies on flake registry
					exec("nix", "shell", "nixpkgs#" . @$res[$choice - 1]->{package},
						"--command", shell_quote(@ARGV));
				} else {
					exec("nix-shell", "-p", @$res[$choice - 1]->{package},
						"--run", shell_quote("exec", @ARGV));
				}
            }
        }
    } else {
        print STDERR <<EOF;
The program '$program' is not in your PATH. It is provided by several packages.
You can make it available in an ephemeral shell by typing one of the following:
EOF
        print STDERR "  nix-shell -p $_->{package}\n" foreach @$res;
    }
}

exit 127;
